### PR TITLE
Use os_sys_calls in all code paths to replace raw syscalls

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Description:
- This PR seeks to address remaining changes for cross-platform code (ultimately for Windows support) that apply to the linux and os/x code paths in order to consolidate the necessary patch review requirements.
- The subsequent PR(s) should largely avoid changes impacting the current working linux and os/x code paths and should be targeted mainly at Windows code.
- Primarily adds usage of os_sys_calls instead of directly using syscalls for various network and test sources.
- Addresses a reoccurring issue that only linux/posix allows for connection to an any-address. On Windows and in general, this makes no sense so several tests are adjusted to explicitly connect to the loopback address instead of the "any" address.
- Includes changes to simplify platform specific includes/macros
- Coalesces between platforms (posix/Windows) and hopefully simplifies the implementation of io_socket_handle_impl
Risk Level: Low
Testing: Local Linux gcc and Windows MSVC
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: test/common/network/connection_impl_test.cc [hunk 1]
```diff
@@ -1991,8 +1997,11 @@ class ReadBufferLimitTest : public ConnectionImplTest {
     listener_ = dispatcher_->createListener(socket_, listener_callbacks_, true);
 
     client_connection_ = dispatcher_->createClientConnection(
-        socket_->localAddress(), Network::Address::InstanceConstSharedPtr(),
-        Network::Test::createRawBufferSocket(), nullptr);
+        Utility::resolveUrl(fmt::format("tcp://{}:{}",
+                                        Network::Test::getLoopbackAddressUrlString(GetParam()),
+                                        socket_->localAddress()->ip()->port())),
```
Comment (zuercher) [inline test/common/network/connection_impl_test.cc]: Does `socket_->localAddress()` on Win32 not return the listener address selected? Seems like we might depend on that elsewhere.

File: test/common/network/listen_socket_impl_test.cc [hunk 2]
```diff
@@ -94,11 +96,6 @@ class ListenSocketImplTest : public testing::TestWithParam<Address::IpVersion> {
       // The address and port are bound already, should throw exception.
       EXPECT_THROW(createListenSocketPtr(addr, options2, true), SocketBindException);
 
-      // Test the case of a socket with fd and given address and port.
-      IoHandlePtr dup_handle = std::make_unique<IoSocketHandleImpl>(dup(socket1->ioHandle().fd()));
-      auto socket3 = createListenSocketPtr(std::move(dup_handle), addr, nullptr);
-      EXPECT_EQ(addr->asString(), socket3->localAddress()->asString());
-
```
Comment (zuercher) [inline test/common/network/listen_socket_impl_test.cc]: Should this just be disabled for Win32?

File: test/common/network/listener_impl_test.cc [hunk 3]
```diff
@@ -228,9 +228,12 @@ TEST_P(ListenerImplTest, DisableAndEnableListener) {
...
 
-  ClientConnectionPtr client_connection =
-      dispatcher_->createClientConnection(socket->localAddress(), Address::InstanceConstSharedPtr(),
-                                          Network::Test::createRawBufferSocket(), nullptr);
+  Network::Address::InstanceConstSharedPtr remote = Utility::resolveUrl(
+      fmt::format("tcp://{}:{}", Network::Test::getLoopbackAddressUrlString(GetParam()),
+                  socket->localAddress()->ip()->port()));
+
+  ClientConnectionPtr client_connection = dispatcher_->createClientConnection(
+      remote, Address::InstanceConstSharedPtr(), Network::Test::createRawBufferSocket(), nullptr);
```
Comment (zuercher) [inline test/common/network/listener_impl_test.cc]: Same question as in connection_impl_test.cc.

File: source/common/network/address_impl.cc [hunk 4]
```diff
@@ -398,16 +425,18 @@ PipeInstance::PipeInstance(const std::string& pipe_path, mode_t mode) : Instance
...
 
 Api::SysCallIntResult PipeInstance::bind(os_fd_t fd) const {
-  if (!abstract_namespace_) {
-    // Try to unlink an existing filesystem object at the requested path. Ignore
-    // errors -- it's fine if the path doesn't exist, and if it exists but can't
-    // be unlinked then `::bind()` will generate a reasonable errno.
-    unlink(address_.sun_path);
+  Api::OsSysCalls& os_sys_calls = Api::OsSysCallsSingleton::get();
+  if (abstract_namespace_) {
+    return os_sys_calls.bind(fd, reinterpret_cast<const sockaddr*>(&address_),
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/network/address_impl.cc]: Agreed, there is a much simpler way following this code pattern. I believe the change dates to long before the cleaner sockaddr accessors, since windows API uses a number of incomplete structs for these things, and the sizeof sockaddr_storage is generally preferred (except in this case.) Fixing in the next refresh.

File: source/common/network/address_impl.cc [hunk 5]
```diff
@@ -416,8 +445,12 @@ Api::SysCallIntResult PipeInstance::bind(os_fd_t fd) const {
...
 
 Api::SysCallIntResult PipeInstance::connect(os_fd_t fd) const {
-  const int rc = ::connect(fd, sockAddr(), sockAddrLen());
-  return {rc, errno};
+  Api::OsSysCalls& os_sys_calls = Api::OsSysCallsSingleton::get();
+  if (abstract_namespace_) {
+    return os_sys_calls.connect(fd, reinterpret_cast<const sockaddr*>(&address_),
+                                offsetof(struct sockaddr_un, sun_path) + address_length_);
+  }
+  return os_sys_calls.connect(fd, reinterpret_cast<const sockaddr*>(&address_), sizeof(address_));
```
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/network/address_impl.cc]: Yes, same and simpler.

File: test/common/network/connection_impl_test.cc (hunk 1 above)
Comment (wrowe) [reply test/common/network/connection_impl_test.cc]: For the "any" interface? Interesting question, I'll research.

File: test/common/network/listen_socket_impl_test.cc (hunk 2 above)
Comment (wrowe) [reply test/common/network/listen_socket_impl_test.cc]: No. That test logic is a complete no-op, our test goal is not to confirm that linux ::dup() works. The result is never used.

File: test/common/network/listener_impl_test.cc (hunk 3 above)
Comment (wrowe) [reply test/common/network/listener_impl_test.cc]: Yes. Researching.

File: test/common/network/listen_socket_impl_test.cc (hunk 2 above)
Comment (wrowe) [reply test/common/network/listen_socket_impl_test.cc]: (We had the same initial reaction, but as we looked at that test pattern, there appears to be no purpose and no reason to implement os_sys_calls.dup() at this time.)

Comment (zuercher) [reply test/common/network/listen_socket_impl_test.cc]: Isn't this testing the `Envoy::Network::NetworkListenSocket` constructor that takes an `IoHandlePtr` vs. `Address::InstanceConstSharedPtr`. There is probably another way to test that without using `::dup()`, though.

File: source/common/network/address_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/network/address_impl.cc]: The next commit simplifies this change, thanks!

File: source/common/network/address_impl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wrowe) [reply source/common/network/address_impl.cc]: You were correct, the next commit simplifies this change.

File: test/common/network/connection_impl_test.cc (hunk 1 above)
Comment (wrowe) [reply test/common/network/connection_impl_test.cc]: It appears this works (and if problems are encountered, localAddress() needs fixing.) Adopted the proposed change for this next commit.

File: test/common/network/listener_impl_test.cc (hunk 3 above)
Comment (wrowe) [reply test/common/network/listener_impl_test.cc]: Adopted, thanks.

File: test/common/network/listen_socket_impl_test.cc (hunk 2 above)
Comment (sunjayBhatia) [reply test/common/network/listen_socket_impl_test.cc]: we addressed this by removing dup and closing the fd before creating a new socket and using that to construct an `IoHandlePtr`

**Comment by sunjayBhatia** [comment]
Circle CI failing due to https://github.com/envoyproxy/envoy/issues/10012

Only `//test/extensions/quic_listeners/quiche/integration:quic_http_integration_test` fails locally for us

File: test/common/network/listen_socket_impl_test.cc (hunk 2 above)
Comment (sunjayBhatia) [reply test/common/network/listen_socket_impl_test.cc]: resolving as we addressed the removal of `dup`

File: source/common/network/address_impl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply source/common/network/address_impl.cc]: resolving, we re-simplified this code

File: source/common/network/address_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply source/common/network/address_impl.cc]: resolving, we re-simplified this code

File: source/common/network/address_impl.cc [hunk 6]
```diff
@@ -230,8 +255,8 @@ Api::SysCallIntResult Ipv4Instance::bind(os_fd_t fd) const {
 }
 
 Api::SysCallIntResult Ipv4Instance::connect(os_fd_t fd) const {
-  const int rc = ::connect(fd, sockAddr(), sockAddrLen());
-  return {rc, errno};
+  return Api::OsSysCallsSingleton::get().connect(
+      fd, reinterpret_cast<const sockaddr*>(&ip_.ipv4_.address_), sizeof(ip_.ipv4_.address_));
```
Comment (zuercher) [inline source/common/network/address_impl.cc]: Can we use sockAddr() and sockAddrLen() here and below?

Comment (sunjayBhatia) [reply source/common/network/address_impl.cc]: changed this, resolving

**Comment by sunjayBhatia** [comment]
@zuercher any idea how to get CI for this unstuck? looks like the coverage tests failed again on the same flake but not even the sign-off check has run

**Comment by wrowe** [comment]
Confirming CircleCI coverage is failing on the usual flake (same as issue #10012), but this PR appears good to go;

[ RUN      ] Protocols/BufferIntegrationTest.RouterRequestPopulateContentLengthOnTrailers/IPv4_HttpDownstream_HttpUpstream
external/bazel_tools/tools/test/collect_coverage.sh: line 128:  1002 Segmentation fault      (core dumped) "$@"
+ TEST_STATUS=139
+ touch /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/4629/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/coverage/coverage_tests/shard_27_of_50/coverage.dat
+ [[ 139 -ne 0 ]]
+ echo --
--
+ echo Coverage runner: Not collecting coverage for failed test.

**Comment by zuercher** [comment]
@mattklein123 (or maybe @lizan) I think if this is to be merged without coverage passing one of you is going to have to push the button.

File: source/common/network/connection_impl.cc [hunk 7]
```diff
@@ -205,6 +212,10 @@ void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
   connection_stats_.reset();
 
   file_event_.reset();
+
+  // Shutdown the socket so the other side can read any data in flight.
+  Api::OsSysCallsSingleton::get().shutdown(ioHandle().fd(), ENVOY_SHUT_WR);
```
Comment (mattklein123) [inline source/common/network/connection_impl.cc]: This is a functional change on Linux, right? Can you revert this and we can do an independent PR on this and discuss?

**Comment by lizan** [comment]
Can you merge master to fix coverage?

**Comment by sunjayBhatia** [comment]
> Can you merge master to fix coverage?

will do!

File: source/common/network/connection_impl.cc (hunk 7 above)
Comment (wrowe) [reply source/common/network/connection_impl.cc]: Agreed, thanks @mattklein123 , addressed and then remerging master to possibly pass coverage.
